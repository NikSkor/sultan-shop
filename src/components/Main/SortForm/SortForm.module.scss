.form {
  width: 100%;
  display: flex;
  flex-direction: column;
}

.title {
  font-weight: 500;
  font-size: 16px;
  line-height: 150%;
  color: #111111;
  margin-top: 20px;
  margin-bottom: 10px;
}

.price {
  display: flex;
  gap: 5px;
  font-weight: 300;
  font-size: 14px;
  line-height: 150%;
  color: #3F4E65;
  margin-bottom: 16px;
}

.currency {
  font-weight: 500;
  font-size: 14px;
  line-height: 150%;
  color: #111111;
}

.priceRangeBlock {
  width: 238px;
  display: flex;
  gap: 10px;
  border-bottom: 1px dashed rgba(63, 78, 101, .3);
  padding-bottom: 30px;
}

.inputNumber {
  min-width: 100px;
  appearance: none;
  padding: 4px 5px;
  font-weight: 300;
  font-size: 12px;
  line-height: 15px;
  text-align: center;
  color: #3F4E65;
  border: none;
  outline: none;
  background-image: linear-gradient(90deg, rgba(255, 198, 80, 0.3) 0%, rgba(254, 202, 110, 0.3) 97.25%);
  border-radius: 1000px;
}

.symbol {
  font-weight: 500;
  font-size: 14px;
  line-height: 130%;
  color: #3F4E65;
}

.brandSort {
  width: 238px;
}

.checkboxBlock {
  margin-top: 15px;
  display: flex;
  flex-direction: column;
  margin-bottom: 25px;
}

.checkboxLabel {
  display: flex;
  gap: 10px;
  align-items: center;
  font-weight: 400;
  font-size: 14px;
  line-height: 150%;
  color: #3F4E65;
}

.submitBlock {
  display: flex;
  gap: 10px;
  margin-bottom: 40px;
}

.showBtn {
  background-color: #FFC85E;
  border-radius: 79px;
  padding: 21px 51px;
  font-weight: 700;
  font-size: 14px;
  line-height: 17px;
  color: #FFFFFF;
}

.cleaneBtn {
  background: #FFC85E;
  border-radius: 79px;
  color: #FFFFFF;
  padding: 10px 19px;
}