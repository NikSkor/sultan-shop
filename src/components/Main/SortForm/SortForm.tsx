import React, { FC } from 'react';
import Search from '../../Header/Search/Search';
import style from './SortForm.module.scss';
import trashImg from '../../../img/trash.svg';
import { useAppDispatch, useAppSelector } from '../../../hooks/redux';
import { userSlice } from '../../../store/reducers/UserSlice';

const SortForm: FC = () => {
  
  const brands = useAppSelector(state => state.userReducer.search);
  const brandsFilter = useAppSelector(state => state.userReducer.brand);
  const minPriceInput = useAppSelector(state => state.userReducer.minPriceInput);
  const maxPriceInput = useAppSelector(state => state.userReducer.maxPriceInput);
  const brandList = useAppSelector(state => state.userReducer.brandArr);

  let brandArr: string[] = [...brandList];

  const filterBrand = (brands: string[], name: string) => {
    if(name === '') return brands;

    let searchBrands: string[] = [];

      brands.forEach((item: string) => {
      if(item.toLowerCase().includes(name.toLocaleLowerCase())) {
        searchBrands.push(item);
      }
    });
    return searchBrands;
  }

  brandArr = filterBrand(brandArr, brands);

  const {filterBrands} = userSlice.actions;
  const {filterCatalogBrands} = userSlice.actions;
  const {clearBrands} = userSlice.actions;
  const {clearSearch} = userSlice.actions;
  const {inputMinPrice} = userSlice.actions;
  const {inputMaxPrice} = userSlice.actions;
  const {minPrice} = userSlice.actions;
  const {maxPrice} = userSlice.actions;

  const dispatch = useAppDispatch();

  return (
    <div className={style.form}>
      <div className={style.priceSort}>
        <h3 className={style.title}>ПОДБОР ПО ПАРАМЕТРАМ</h3>
        <div className={style.price}>Цена<p className={style.currency}>₸</p></div>
        <div className={style.priceRangeBlock}>
          <input 
            data-testid = 'testNumberFrom'
            type="number" 
            className={style.inputNumber}
            value = {minPriceInput}
            onChange={(e) => {
              dispatch(inputMinPrice(e.target.value));
            }}
            />
          <p className={style.simbol}>-</p>
          <input 
            data-testid = 'testNumberTo'
            type="number" 
            className={style.inputNumber}
            value = {maxPriceInput}
            onChange={(e) => {
              dispatch(inputMaxPrice(e.target.value));
            }}
            />
        </div>
      </div>
      <div className={style.brandSort}>
        <h3 className={style.title} style={{marginBottom: '15px'}}>Бренд</h3>
        <Search/>
        <div className={style.checkboxBlock}>
          {brandArr.map((item) => {
            return (
              <label key={item} className={style.checkboxLabel}>
                <input 
                type="checkbox"
                data-id = {item}
                checked = {
                  (brandsFilter.includes(item))
                  ? true
                  : false
                }
                onChange = {(e)=> {
                  let str: string | undefined = e.target.dataset['id'];
                  
                  if(typeof str === 'string') {
                    dispatch(filterBrands(str));
                  }
                }}
                />
                  {item}
              </label>
              ) 
          })}
        </div>
        <div className={style.submitBlock}>
          <button 
            className={style.showBtn}
            onClick={(e)=> {
              e.preventDefault();
              if (brandsFilter.length !== 0){
                dispatch(filterCatalogBrands());
              };
              
              dispatch(minPrice());
              dispatch(maxPrice());
            }}
            >Показать</button>
          <button 
            className={style.cleaneBtn}
            onClick={(e)=> {
              dispatch(clearBrands());
              dispatch(clearSearch());
            }}
            >
            <img src={trashImg} alt="Значок очистки" />
          </button>
        </div>

      </div>
    </div>
  )
}

export default SortForm;