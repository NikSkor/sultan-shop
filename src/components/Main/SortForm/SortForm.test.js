import { fireEvent, render, screen } from '@testing-library/react';
import { Provider } from 'react-redux';
import { MemoryRouter } from 'react-router-dom';
import { setupStore } from '../../../store/store.ts';
import SortForm from './SortForm.tsx';

const store = setupStore();

describe('SortForm component, inputs price from', () => {
  test('number input price from correctly completed', () => {
    render(
      <MemoryRouter>
        <Provider store={store}>
          <SortForm />
        </Provider>
      </MemoryRouter>
    );
    const input = screen.getByTestId('testNumberFrom');
    expect(input).toContainHTML('');
    fireEvent.input(input, {
      target: { value: '300' },
    });
    expect(input).toContainHTML('300');
  });
  test('number input price from not correctly completed', () => {
    render(
      <MemoryRouter>
        <Provider store={store}>
          <SortForm />
        </Provider>
      </MemoryRouter>
    );
    const input = screen.getByTestId('testNumberFrom');
    expect(input).toContainHTML('');
    fireEvent.input(input, {
      target: { value: 'sjdf' },
    });
    expect(input).not.toContainHTML('300');
  });
  test('number input price from is full', () => {
    render(
      <MemoryRouter>
        <Provider store={store}>
          <SortForm />
        </Provider>
      </MemoryRouter>
    );
    const input = screen.getByTestId('testNumberFrom');
    expect(input).toContainHTML('');
    fireEvent.input(input, {
      target: { value: '' },
    });
    expect(input).toContainHTML('');
  });
});

describe('SortForm component, inputs price to', () => {
  test('number input price to correctly completed', () => {
    render(
      <MemoryRouter>
        <Provider store={store}>
          <SortForm />
        </Provider>
      </MemoryRouter>
    );
    const input = screen.getByTestId('testNumberTo');
    expect(input).toContainHTML('');
    fireEvent.input(input, {
      target: { value: '900' },
    });
    expect(input).toContainHTML('900');
  });
  test('number input price to not correctly completed', () => {
    render(
      <MemoryRouter>
        <Provider store={store}>
          <SortForm />
        </Provider>
      </MemoryRouter>
    );
    const input = screen.getByTestId('testNumberTo');
    expect(input).toContainHTML('');
    fireEvent.input(input, {
      target: { value: 'sjdf' },
    });
    expect(input).not.toContainHTML('900');
  });
  test('number input price to is full', () => {
    render(
      <MemoryRouter>
        <Provider store={store}>
          <SortForm />
        </Provider>
      </MemoryRouter>
    );
    const input = screen.getByTestId('testNumberTo');
    expect(input).toContainHTML('');
    fireEvent.input(input, {
      target: { value: '' },
    });
    expect(input).toContainHTML('');
  });
});

describe('SortForm component, buttons', () => {
  test('buttons are correctly completed', () => {
    render(
      <MemoryRouter>
        <Provider store={store}>
          <SortForm />
        </Provider>
      </MemoryRouter>
    );

    const btns = screen.getAllByRole('button');
    btns.forEach((item)=> {
      expect(item).toBeInTheDocument();
      expect(item).toMatchSnapshot();
    })

  });
});
