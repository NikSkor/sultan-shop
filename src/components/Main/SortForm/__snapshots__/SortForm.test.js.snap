// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SortForm component, buttons buttons are correctly completed 1`] = `
<button
  class="btn"
>
  <svg
    class="svg"
    fill="none"
    height="39"
    viewBox="0 0 39 39"
    width="39"
    xmlns="http://www.w3.org/2000/svg"
  >
    <rect
      class="rect"
      fill="currentColor"
      height="39"
      rx="19.5"
      width="39"
    />
    <path
      class="path"
      d="M27.5297 27.5294L24.0992 24.0928L27.5297 27.5294ZM26.0002 19.5C26.0002 21.2239 25.3154 22.8772 24.0964 24.0962C22.8775 25.3152 21.2242 26 19.5002 26C17.7763 26 16.123 25.3152 14.9041 24.0962C13.6851 22.8772 13.0002 21.2239 13.0002 19.5C13.0002 17.7761 13.6851 16.1228 14.9041 14.9038C16.123 13.6848 17.7763 13 19.5002 13C21.2242 13 22.8775 13.6848 24.0964 14.9038C25.3154 16.1228 26.0002 17.7761 26.0002 19.5V19.5Z"
      stroke="currentColor"
      stroke-linecap="round"
      stroke-width="1.3"
    />
  </svg>
</button>
`;

exports[`SortForm component, buttons buttons are correctly completed 2`] = `
<button
  class="showBtn"
>
  Показать
</button>
`;

exports[`SortForm component, buttons buttons are correctly completed 3`] = `
<button
  class="cleaneBtn"
>
  <img
    alt="Значок очистки"
    src="trash.svg"
  />
</button>
`;

exports[`SortForm component, checkbox checkbox correctly completed 1`] = `
<button
  class="btn"
>
  <svg
    class="svg"
    fill="none"
    height="39"
    viewBox="0 0 39 39"
    width="39"
    xmlns="http://www.w3.org/2000/svg"
  >
    <rect
      class="rect"
      fill="currentColor"
      height="39"
      rx="19.5"
      width="39"
    />
    <path
      class="path"
      d="M27.5297 27.5294L24.0992 24.0928L27.5297 27.5294ZM26.0002 19.5C26.0002 21.2239 25.3154 22.8772 24.0964 24.0962C22.8775 25.3152 21.2242 26 19.5002 26C17.7763 26 16.123 25.3152 14.9041 24.0962C13.6851 22.8772 13.0002 21.2239 13.0002 19.5C13.0002 17.7761 13.6851 16.1228 14.9041 14.9038C16.123 13.6848 17.7763 13 19.5002 13C21.2242 13 22.8775 13.6848 24.0964 14.9038C25.3154 16.1228 26.0002 17.7761 26.0002 19.5V19.5Z"
      stroke="currentColor"
      stroke-linecap="round"
      stroke-width="1.3"
    />
  </svg>
</button>
`;

exports[`SortForm component, checkbox checkbox correctly completed 2`] = `
<button
  class="showBtn"
>
  Показать
</button>
`;

exports[`SortForm component, checkbox checkbox correctly completed 3`] = `
<button
  class="cleaneBtn"
>
  <img
    alt="Значок очистки"
    src="trash.svg"
  />
</button>
`;
